pkg/querier/queryrange/stats_test.go

```diff
@@ -0,0 +1,103 @@
+)
+
+func TestStatsMiddleware(t *testing.T) {
+	md := StatsMiddleware()
+	for name, test := range map[string]struct {
+		queryrange.Handler
+		queryrange.Request
+		expect func(t *testing.T, status, query string, rangeType logql.QueryRangeType, stats stats.Result)
+	}{
+		"Prometheus instant query nil": {
+			Handler: queryrange.HandlerFunc(func(c context.Context, r queryrange.Request) (queryrange.Response, error) {
+				return &LokiPromResponse{}, nil
+			}),
+			Request: &LokiRequest{
+				Query: "foo",
+			},
+			expect: func(t *testing.T, status, query string, rangeType logql.QueryRangeType, stats stats.Result) {
+				require.Equal(t, "", status)
+				require.Equal(t, "foo", query)
+				require.Equal(t, logql.InstantType, rangeType)
+				require.NotEmpty(t, stats.Summary.ExecTime)
+			},
+		},
+		"Prometheus instant query": {
+			Handler: queryrange.HandlerFunc(func(c context.Context, r queryrange.Request) (queryrange.Response, error) {
+				return &LokiPromResponse{
+					Response: &queryrange.PrometheusResponse{
+						Status: loghttp.QueryStatusSuccess,
+					},
+				}, nil
+			}),
+			Request: &LokiRequest{
+				Query: "foo",
+			},
+			expect: func(t *testing.T, status, query string, rangeType logql.QueryRangeType, stats stats.Result) {
+				require.Equal(t, loghttp.QueryStatusSuccess, status)
+				require.Equal(t, "foo", query)
+				require.Equal(t, logql.InstantType, rangeType)
+				require.NotEmpty(t, stats.Summary.ExecTime)
+			},
+		},
+		"Loki range query": {
+			Handler: queryrange.HandlerFunc(func(c context.Context, r queryrange.Request) (queryrange.Response, error) {
+				return &LokiResponse{
+					Status: loghttp.QueryStatusFail,
+					Statistics: stats.Result{
+						Store: stats.Store{
+							DecompressedBytes: 20 * 1024,
+						},
+					},
+				}, nil
+			}),
+			Request: &LokiRequest{
+				Query: "foo",
+				EndTs: time.Now(),
+			},
+			expect: func(t *testing.T, status, query string, rangeType logql.QueryRangeType, stats stats.Result) {
+				require.Equal(t, loghttp.QueryStatusFail, status)
+				require.Equal(t, "foo", query)
+				require.Equal(t, logql.RangeType, rangeType)
+				require.NotEmpty(t, stats.Summary.ExecTime)
+				require.Equal(t, int64(20*1024), stats.Store.DecompressedBytes)
+			},
+		},
+		"error": {
+			Handler: queryrange.HandlerFunc(func(c context.Context, r queryrange.Request) (queryrange.Response, error) {
+				return nil, errors.New("")
+			}),
+			Request: &LokiRequest{
+				Query: "foo",
+				EndTs: time.Now(),
+			},
+			expect: func(t *testing.T, status, query string, rangeType logql.QueryRangeType, stats stats.Result) {
+				require.Equal(t, loghttp.QueryStatusFail, status)
+				require.Equal(t, "foo", query)
+				require.Equal(t, logql.RangeType, rangeType)
+				require.NotEmpty(t, stats.Summary.ExecTime)
+			},
+		},
+	} {
+		t.Run(name, func(t *testing.T) {
+			RecordMetrics = func(status, query string, rangeType logql.QueryRangeType, stats stats.Result) {
```

Comment:
This approach definitely works, but this kind of package level variable overriding makes me very nervous as opposed to a struct middleware which has a field `RecordMetrics`.

Reply:
Yeah I agree, I tried to avoid having a complex object to build for this. But yeah ,might be cleaner to use a struct. I'll look into it.